lib/openstack_query/runners/server_runner.py

```diff
@@ -0,0 +1,104 @@
+from typing import Optional, Dict, List
+
+from openstack.compute.v2.server import Server
+from openstack.identity.v3.project import Project
+
+from openstack_api.openstack_connection import OpenstackConnection
+from openstack_query.runners.query_runner import QueryRunner
+
+from exceptions.parse_query_error import ParseQueryError
+
+
+class ServerRunner(QueryRunner):
+    """
+    Runner class for openstack Server resource.
+    ServerRunner encapsulates running any openstacksdk Server commands
+    """
+
+    def _run_query(
+        self,
+        conn: OpenstackConnection,
+        filter_kwargs: Optional[Dict[str, str]] = None,
+        **kwargs
+    ) -> List[Server]:
+        """
+        This method runs the query by running openstacksdk commands
+
+        For QueryServer, this command gets all projects available and iteratively finds servers that belong to that project
+        :param conn: An OpenstackConnection object - used to connect to openstacksdk
+        :param filter_kwargs: An Optional set of filter kwargs to pass to conn.compute.servers()
+            to limit the servers being returned. - see https://docs.openstack.org/api-ref/compute/#list-servers
+
+        :param kwargs: a set of extra meta params that override default openstacksdk command. Current valid kwargs are:
+
+            'from_projects': Union[List[str], List[Project]] - takes a list of projects to run the query on
+
+            'from_subset': Union[List[str], List[Server]] - takes a list of servers to run the query on
+            (forces use of filter functions)
+
+        """
+        projects = []
+        if "from_projects" in kwargs:
+            if len(kwargs["from_projects"]) == 0:
+                raise ParseQueryError(
+                    "'from_project' meta kwarg given but no projects given to search in"
+                )
+            for project in kwargs["from_projects"]:
+                if not isinstance(project, Project):
+                    raise ParseQueryError(
+                        "'from_project' only accepts Openstack 'Project' objects"
+                    )
+                projects.append(project)
```

Comment:
This should be extracted into a private method and just assign the result
```
projects = self._get_projects(kwargs.get("from_projects", None))

# ...
ProjectIdentifier = Union[str, Project]
def _get_projects(projects: Optional[ProjectIdentifier]) -> return List[ProjectIdentifier]:
    pass

Reply:
I would argue in a future PR we'll want to extract this as it's something that will apply for everything

Reply:
You mentioned previously that openstack `find_server` is slow - I need to use it here to get each project from its identifier - I think in this case its worthwhile to do since you mentioned that `from_project` would apply for everything